packages/components/src/unit-control/utils.js

```diff
@@ -138,6 +237,48 @@ export function getValidParsedUnit( next, units, fallbackValue, fallbackUnit ) {
  * @return {string} a11y label for the unit abbreviation
  */
 export function parseA11yLabelForUnit( unit ) {
-	const match = CSS_UNITS.find( ( item ) => item.value === unit );
+	const match = ALL_CSS_UNITS.find( ( item ) => item.value === unit );
 	return match?.a11yLabel ? match?.a11yLabel : match?.value;
 }
+
+/**
+ * Filters available units based on values defined by settings.
+ *
+ * @param {Array} settings Collection of preferred units.
+ * @param {Array} units Collection of available units.
+ *
+ * @return {Array} Filtered units based on settings.
+ */
+function filterUnitsWithSettings( settings = [], units = [] ) {
+	return units.filter( ( unit ) => {
+		return settings.includes( unit.value );
+	} );
+}
+
+/**
+ * Custom hook to retrieve and consolidate units setting from add_theme_support().
+ *
+ * @param {Object} args                An object containing units, settingPath & defaultUnits.
+ * @param {Object} args.units          Collection of available units.
+ * @param {string} args.availableUnits The setting path. Defaults to 'spacing.units'.
+ * @param {Object} args.defaultValues  Collection of default values for defined units. Example: { px: '350', em: '15' }.
+ *
+ * @return {Array} Filtered units based on settings.
+ */
+export const useCustomUnits = ( { units, availableUnits, defaultValues } ) => {
```

Comment:
Thanks for the PR @aristath. Thinking a bit I'm not at ease with exposing this hook in the components package. Can you clarify how it's supposed to be used? Why do we need both this hook and the component with the units prop. 

I guess on my mind doing something like this should be always sufficient,  why it's not the case?

```
<UnitControl units={ useSetting( 'spacing.units' ) } />
````

Reply:
Yeah this was a very deep rabbit hole and I spent hours trying and iterating things to find something that works... :sweat_smile: 

These controls are used in lots of places, in different contexts. The default values vary depending on the context (for example the cover block's height should by default include `vh` units while the layout or column-width units should not).

`useSetting` will only get the value from `theme.json` which is formatted like `["px", "em", "rem"]`. The control however requires a different format so `useCustomUnits` needs to run in order to convert the units to `[{value:"em", ...}, {...}, ...]` (that's what we were doing before too).
I suppose we could make `UnitControl` understand an array of units like `["px","em","rem"]` and just make the conversion to the full array inside the control itself... But then if `spacing.units` is empty it would not fallback to the default units as it does now.
Using `useSetting` to get these would not work correctly unless we add hardcoded fallback in `useSetting` to get defaults - both in the block-editor and the site-editor packages since `useSetting` is implemented separately in each of them (which is also what triggered the experimentation on https://github.com/WordPress/gutenberg/pull/31782).

The reason I placed it in `components` inside the control's utils is because it seemed that the most suitable place for it... Do you think some other package would be a better home?

Reply:
> I suppose we could make UnitControl understand an array of units like ["px","em","rem"] and just make the conversion to the full array inside the control itself... But then if spacing.units is empty it would not fallback to the default units as it does now.

I feel this might be the best way potentially but right now maybe we can just start by marking the hook as experimental?

Reply:
I went ahead and marked the hook as experimental for now.